src/Processors/Formats/Impl/Parquet/ParquetBloomFilterCondition.cpp

```diff
@@ -0,0 +1,497 @@
+
+                bool column_maybe_contains = maybeTrueOnBloomFilter(
+                    element.hashes_per_column[column_index],
+                    column_index_to_column_bf.at(element.key_columns[column_index]));
+
+                if (!column_maybe_contains)
+                {
+                    maybe_true = false;
+                    break;
+                }
+            }
+
+            rpn_stack.emplace_back(maybe_true, true);
+            if (element.function == Function::FUNCTION_NOT_IN)
+                rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_NOT)
+        {
+            rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_OR)
+        {
+            auto arg1 = rpn_stack.back();
+            rpn_stack.pop_back();
+            auto arg2 = rpn_stack.back();
+            rpn_stack.back() = arg1 | arg2;
+        }
+        else if (element.function == Function::FUNCTION_AND)
+        {
+            auto arg1 = rpn_stack.back();
+            rpn_stack.pop_back();
+            auto arg2 = rpn_stack.back();
+            rpn_stack.back() = arg1 & arg2;
+        }
+        else if (element.function == Function::ALWAYS_TRUE)
+        {
+            rpn_stack.emplace_back(true, false);
+        }
+        else
+        {
+            rpn_stack.emplace_back(true, true);
+        }
+    }
+
+    if (rpn_stack.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected stack size in KeyCondition::mayBeTrueOnRowGroup");
+
+    return rpn_stack[0].can_be_true;
+}
+
+std::unordered_set<std::size_t> ParquetBloomFilterCondition::getFilteringColumnKeys() const
+{
+    std::unordered_set<std::size_t> column_keys;
+
+    for (const auto & element : condition)
+    {
+        for (const auto index : element.key_columns)
+        {
+            column_keys.insert(index);
+        }
+    }
+
+    return column_keys;
+}
+
+/*
+ * `KeyCondition::rpn` is overly complex for bloom filters, some operations are not even supported. Not only that, but to avoid hashing each time
+ * we loop over a rpn element, we need to store hashes instead of where predicate values. To address this, we loop over `KeyCondition::rpn`
+ * and build a simplified RPN that holds hashes instead of values.
+ *
+ * `KeyCondition::RPNElement::FUNCTION_IN_RANGE` becomes:
+ *      `FUNCTION_EQUALS` when range limits are equal
+ *      `FUNCTION_UNKNOWN` when range limits are different
+ * `KeyCondition::RPNElement::FUNCTION_IN_SET` becomes
+ *      `FUNCTION_IN`
+ *
+ * Complex types and structs are not supported.
+ * There are two sources of data types being analyzed, and they need to be compatible: DB::Field type and parquet type.
+ * This is determined by the `isColumnSupported` method.
+ *
+ * Some interesting examples:
+ * 1. file(..., 'string_column UInt64') where str_column = '50'; Field.type == UInt64. Parquet type string. Not supported.
```

Comment:
I am sorry, I am a bit confused by this comment. PLus, what do you mean by `s/where/containing`?

For the first line, I am trying to showcase a string column being read as uint64 and then compred to a string value.

Reply:
Ok, I misunderstood it then; the "where" is meant as part of SQL query after all.

Isn't `str_column = '50'` effectively the same as `str_column = 50` as far as this code is concerned, because KeyCondition would convert the constant to UInt64?

What confused me is the comparison between UInt64 clickhouse column and string constant. Since `Field.type == UInt64` the string `'50'` in the query seems like an unnecessary complication. So I thought maybe the `where str_column = '50'` part describes file contents (where '50' would actually be a string) rather than the query.

I guess the example was meant to demonstrate two points at once: that bloom filter is not used when parquet file has string and ch has integer, and that KeyCondition converts constant to the type of key column. I'd split it into two examples and explicitly mention that KeyCondition converts constants to types of key columns.

```suggestion
 * 1. file(..., 'str_column UInt64') where str_column = 50; Field.type == UInt64. Parquet type string. Not supported.
```

Reply:
Yes, I meant to show that a String column can be read as a UInt64 and then can be compared to a String. The two conversions that can take place.

In any case, I simplified it a bit for the sake of bloom filters. Second row already shows UInt64 can be compared to String and that conversion happens somewhere else.